### PR TITLE
WIP/RFC: Iteration API

### PR DESCRIPTION
**Pull request by shashi**
See #63 

One minor annoyance:

`keys(t::IndexedTable, col::Union{Symbol,Int})` here returns a `Columns` with 1 vector. To get the column, you'd have to do `column(keys(t), col)`... This is so that `keys(t, cols...)`  always works by returning `Columns`. There's an alternative where you could make `keys(t, col::Tuple)` return columns but `keys(t, col::Symbol)` return the vector and not have `keys(t, cols...)`. Not sure which one to go with. @JeffBezanson what do you think?

### CODE REVIEW

File: src/IndexedTables.jl [hunk 1]
```diff
@@ -213,17 +382,17 @@ map{T,D<:Tuple,C<:Tup,V<:Columns}(p::Proj, x::IndexedTable{T,D,C,V}) =
...
+# Given an IndexedTable array with multiple data columns (its data vector is a `Columns` object), return a
+# new array with the specified subset of data columns. Data is shared with the original array.
+# """
+# columns(x::IndexedTable, which...) = IndexedTable(x.index, Columns(x.data.columns[[which...]]), presorted=true)
 
-columns(x::IndexedTable, which) = IndexedTable(x.index, x.data.columns[which], presorted=true)
+#columns(x::IndexedTable, which) = IndexedTable(x.index, x.data.columns[which], presorted=true)
 
-column(x::IndexedTable, which) = columns(x, which)
+#column(x::IndexedTable, which) = columns(x, which)
```
Comment (shashi) [inline src/IndexedTables.jl]: Not sure how to deprecate this...

File: src/utils.jl [hunk 2]
```diff
@@ -358,3 +370,13 @@ Base.@pure @generated function map_params{T<:NamedTuple,S<:NamedTuple}(f, ::Type
     NT = Expr(:macrocall, :(NamedTuples.$(Symbol("@NT"))), fieldnames(T)...)
     :($NT{_map_params(f, T, S)...})
 end
+
+Base.@pure function concat_nt_type{
+           T<:NamedTuple,S<:NamedTuple}(::Type{T}, ::Type{S})
+    namedtuple(fieldnames(T)..., fieldnames(S)...)
+end
```
Comment (shashi) [inline src/utils.jl]: The `@pure` here is unsuccessful at allowing inference for this.

**Comment by shashi** [comment]
I kept the varargs API for getting subsets of columns. I think this PR is in its final form...

**Comment by shashi** [comment]
To make it easier to comment on, here's the change resulted by the alternative API:

https://github.com/JuliaComputing/IndexedTables.jl/commit/e57f0e78a104815d67bd11ef7a3d1850826a1efc

Some awkward things about this:

- `columns(t, :col)` -- is a Vector (the column)
- `rows(t, :col)` -- is also a Vector and is the same as `column(t, :col)`
- `rows(t, :col1, :col2)` is now `rows(t, (:col1, :col2))`
- `rows(t, as(2,:x))` -- gives the vector, so the column has no name
- `rows(t, (as(2, :x),))` -- gives the Columns with the name "x"

But the upside is:

- `keys(t, :col)` and `values(t, :col)` are a Vectors

I'm still ambivalent about which one is good.

**Comment by tkelman** [comment]
`as` and `As` could use docs, not super clear what those are

**Pull request merged**